ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor.hs

```diff
@@ -604,8 +605,9 @@ peerSelectionGovernorLoop tracer
     evalGuardedDecisions :: Time
                          -> PeerSelectionState peeraddr peerconn
                          -> m (TimedDecision m peeraddr peerconn)
-    evalGuardedDecisions blockedAt st =
-      case guardedDecisions blockedAt st of
+    evalGuardedDecisions blockedAt st = do
+      inboundPeers <- readInboundPeers actions
```

Comment:
should this be moved inside `KnownPeers.belowTarget` ?

Reply:
I tried it, but then it's a bit tricky to reason that we don't stale peer sharing when there are no inbound peers; the same RNG might be used when an STM transaction is retried.  I concluded we won't but the argument was subtle, and this is much more explicit, since we have `inboundPeers` when a decision is made.

I made the `readInboundPeers` quite efficient; the value inside of it is computed incrementally so has good amortised bounds; and reading is also cheap (`O(n)` since it's using `Map.map`).

Reply:
We could even make `readInboundPeers` `O(1)` (amortised); we just need to pass `versionData -> DataFlow` to `inbound-governor`.

Reply:
But I thought that maybe in the future, we will need a bit more data from `versionData`, so I left it as is.

Reply:
I see, that makes sense! Thank you for explaining!